File: java/src/main/java/com/voteU/election/java/services/UserService.java
package com.voteU.election.java.services;

import com.voteU.election.java.entities.Role;
import com.voteU.election.java.entities.User;
import com.voteU.election.java.exceptions.ResourceAlreadyExistsException;
import com.voteU.election.java.exceptions.ResourceNotFoundException;
import com.voteU.election.java.repositories.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;
import java.util.Optional;


/**
 * Service class for managing User entities. This class provides methods for
 * creating, retrieving, updating, and deleting user records in the database.
 * It interacts with the UserRepository to facilitate data operations.
 */
@Service
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    /**
     * Retrieves a list of all User entities from the database.
     *
     * @return a list of User objects representing all users in the system.
     */
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    /**
     * Creates a new user in the system. The method sets the current timestamp as the creation time
     * for the user and saves the user entity to the database.
     *
     * @param user the User object to be created and saved in the database
     * @return the created User object after being saved in the database
     */
    public User createUser(User user) {
        if(userRepository.existsByUsername(user.getUsername())) {
            throw new ResourceAlreadyExistsException("Could not create user: " + user.getUsername() + " username already exists");
        }
        if(userRepository.existsByEmail(user.getEmail())) {
            throw new ResourceAlreadyExistsException("Could not create user: " + user.getEmail() + " email already exists");
        }
        user.setCreatedAt(Instant.now());
        String hashedPassword = passwordEncoder.encode(user.getPassword());
        user.setPassword(hashedPassword);
        user.setRole(Role.USER);
        return userRepository.save(user);
    }

    /**
     * Retrieves a user by their unique identifier.
     *
     * @param id the unique identifier of the user to retrieve
     * @return an Optional containing the User if found, or an empty Optional if no user exists with the specified id
     */
    public Optional<User> getUserById(Integer id) {

        if(id == null) {
            throw new ResourceNotFoundException("User not found with id " + id);
        }

        return userRepository.findUserById(id);
    }

    /**
     * Retrieves a user entity based on the provided username.
     *
     * @param username the username of the user to retrieve
     * @return an Optional containing the User entity if found, or an empty Optional if no user was found
     */
    public Optional<User> getUserByUsername(String username) {
        if(username == null) {
            throw new ResourceNotFoundException("User not found with username " + username);
        }

        return (userRepository.findUserByUsername(username));
    }

    /**
     * Retrieves a user from the database based on the provided email.
     *
     * @param email the email address of the user to retrieve
     * @return an Optional containing the User if found, or an empty Optional if no user
     *         exists with the specified email
     */
    public Optional<User> getUserByEmail(String email) {

        if (email == null) {
           throw new ResourceNotFoundException("User not found with email " + email);
        }

        return userRepository.findUserByEmail(email);
    }

    /**
     * Updates a user's information in the system. This method searches for a user based on
     * the provided unique identifier, updates their details with the values provided in the
     * updatedUser object, and saves the changes.
     *
     * @param id the unique identifier of the user to update
     * @param updatedUser the User object containing updated information for the user
     * @return the updated User object after being saved in the database
     * @throws RuntimeException if the user with the given ID is not found
     */
    public User updateUser(Integer id, User updatedUser) {
        return userRepository.findById(id)
                .map(user -> {

                    if (updatedUser.getUsername() != null) {
                        user.setUsername(updatedUser.getUsername());
                    }
                    if (updatedUser.getEmail() != null) {
                        user.setEmail(updatedUser.getEmail());
                    }
                    if (updatedUser.getFirstName() != null) {
                        user.setFirstName(updatedUser.getFirstName());
                    }
                    if (updatedUser.getLastName() != null) {
                        user.setLastName(updatedUser.getLastName());
                    }
                    if (updatedUser.getGender() != null) {
                        user.setGender(updatedUser.getGender());
                    }
                    if (updatedUser.getCountry() != null) {
                        user.setCountry(updatedUser.getCountry());
                    }

                    return userRepository.save(user);
                }).orElseThrow(() -> new ResourceNotFoundException("User not found"));
    }

    /**
     * Deletes a user from the database by their unique identifier.
     *
     * @param id the unique identifier of the user to be deleted
     */
    public void deleteUser(Integer id) {
        userRepository.deleteById(id);
    }


    public User blockUser(Integer id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));
        user.setBlocked(true);
        return userRepository.save(user);
    }

    public User unblockUser(Integer id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));
        user.setBlocked(false);
        return userRepository.save(user);
    }


}

File: voteu-project/src/services/UserService.ts
import { apiFetch } from '@/services/api.ts'
import type { User } from '@/interface/User.ts'

/**
 * Fetches and returns a list of all users.
 *
 * @return {Promise<User[]>} A promise that resolves to an array of User objects.
 */
export async function getAllUsers(): Promise<User[]> {
  return await apiFetch<User[]>('/users');
}

/**
 * Fetches a user by their unique identifier.
 *
 * @param {number} id - The unique identifier of the user to retrieve.
 * @return {Promise<User>} A promise that resolves with the user object.
 */
export async function getUserById(id: number): Promise<User> {
  return await apiFetch<User>(`/users/${id}`);
}

/**
 * Fetches and returns a user by their email address.
 *
 * @param {string} email - The email address of the user to retrieve.
 * @return {Promise<User>} A promise that resolves to the user object.
 */
export async function getUserByEmail(email: string): Promise<User> {
  return await apiFetch<User>(`/users/email/${email}`);
}

/**
 * Retrieves a user by their username.
 *
 * @param {string} username - The username of the user to retrieve.
 * @return {Promise<User>} A promise that resolves to the user object matching the given username.
 */
export async function getUserByUsername(username: string): Promise<User> {
  return await apiFetch<User>(`/users/username/${username}`);
}

/**
 * Creates a new user by sending the provided user details to the API.
 *
 * @param {User} user - The user object containing details of the user to be created.
 * @return {Promise<User>} A promise that resolves to the created user object.
 */
export async function createUser(user: User): Promise<User> {
  return await apiFetch<User>('/users', {
    method: 'POST',
    body: JSON.stringify(user),
  });
}

/**
 * Updates the user information on the server.
 *
 * @param {User} user - An object representing the user to be updated, including its unique identifier and updated properties.
 * @return {Promise<User>} A promise that resolves to the updated user object returned from the server.
 */
export async function updateUser(user: User): Promise<User> {
  return await apiFetch<User>(`/users/${user.id}`, {
    method: 'PUT',
  })
}

/**
 * Deletes a user identified by the given ID.
 *
 * Sends a DELETE request to remove the user associated with the specified ID.
 *
 * @param {number} id - The unique identifier of the user to be deleted.
 * @return {Promise<void>} A promise that resolves when the user is successfully deleted.
 */
export async function deleteUser(id: number): Promise<void> {
  await apiFetch(`/users/${id}`, {
    method: 'DELETE',
    headers: {
      Authorization: `Bearer ${localStorage.getItem('token')}`, // Replace 'localStorage.getItem' as needed in your setup
      'Content-Type': 'application/json',
    },

  })
}

